autoapi/toctree.py

```
@@ -0,0 +1,141 @@
+    """
+
+    if isinstance(node, nodes.section) and \
+            isinstance(node.parent, nodes.document):
+        # Top Level Section header
+        ref_id = docname
+        toc_reference = _find_toc_node(toc, ref_id, nodes.section)
+    elif isinstance(node, nodes.section):
+        # Nested Section header
+        ref_id = node.attributes['ids'][0]
+        toc_reference = _find_toc_node(toc, ref_id, nodes.section)
+    else:
+        # Desc node
+        try:
+            ref_id = node.children[0].attributes['ids'][0]
+            toc_reference = _find_toc_node(toc, ref_id, addnodes.desc)
+        except IndexError as e:
+            print('Invalid desc node: %s' % e)
+            toc_reference = None
+
+    return toc_reference
+
+
+def add_domain_to_toctree(app, doctree, docname):
+    """
+    Add domain objects to the toctree dynamically.
+    This should be attached to the ``doctree-resolved`` event.
+
+    This works by:
+
+    * Finding each domain node (addnodes.desc)
+    * Figuring out it's parent that will be in the toctree
+         (nodes.section, or a previously added addnodes.desc)
+    * Finding that parent in the TOC Tree based on it's ID
+    * Taking that element in the TOC Tree,
+      and finding it's parent that is a TOC Listing (nodes.bullet_list)
+    * Adding the new TOC element for our specific node as a child of that nodes.bullet_list
+        * This checks that bullet_list's last child,
+        and checks that it is also a nodes.bullet_list,
+        effectively nesting it under that element
+
+    """
+
+    toc = app.env.tocs[docname]
+    for desc_node in doctree.traverse(addnodes.desc):
+        try:
+            ref_id = desc_node.children[0].attributes['ids'][0]
+        except IndexError as e:
+            print('Invalid desc node: %s' % e)
+            continue
+        try:
+            # Python domain object
+            ref_text = desc_node[0].attributes['fullname'].split('.')[-1].split('(')[0]
+        except:
```

Bare except here, replace with `KeyError`/`IndexError`